### PR TITLE
Allows Editor to work with our filesystem

### PR DESCRIPTION
**Pull request by twigz20**
This PR:

- Connects the editor to our virtual filesystem.
- Allows the editor to be instantly loaded without having to click on a separate link. [Example](https://twigz20.github.io/next/)

I've filed a couple bugs that need to be solved following this PR.

- #79
- #80
- #81

With this PR I've integrated @humphd 's changes on #75. So I'm closing that PR.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -138,3 +138,6 @@ dist/
 
 # deploy directory
 deploy
+
+#.q files
+q
```
Comment (humphd) [inline .gitignore]: What is this unicode character?

File: package.json [hunk 2]
```diff
@@ -14,8 +14,7 @@
         "build-server": "parcel build src/index.html",
         "build-vm": "node tools/download-vm.js",
         "prebuild-terminal": "npm run build-vm",
-        "build-terminal":
-            "parcel build src/terminal/index.html --public-url /terminal --out-dir dist/terminal",
+        "build-terminal": "parcel build src/terminal/index.html --public-url /terminal --out-dir dist/terminal",
```
Comment (humphd) [inline package.json]: Can you revert this change please?

File: src/editor/Gruntfile.js [hunk 3]
```diff
@@ -591,7 +590,8 @@ module.exports = function(grunt) {
         'requirejs:iframe',
         'exec:localize-dist',
         'build-extensions',
-        'uglify',
+        // 'uglify',
```
Comment (humphd) [inline src/editor/Gruntfile.js]: Do we really need this?  I thought that was just for debugging.

File: src/editor/src/bramble/client/main.js [hunk 4]
```diff
@@ -535,7 +535,7 @@ define([
                 setReadyState(Bramble.MOUNTING);
 
                 root = Path.normalize(root);
-                filename = Path.resolve(root, filename || (_state.fullPath || "index.html"));
+                filename = Path.resolve(root, filename || (_state.fullPath || "editor.html"));
```
Comment (humphd) [inline src/editor/src/bramble/client/main.js]: This change is wrong.  Here we're referring to the `index.html` page in the editor, not for the editor.

File: src/editor/src/routes.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    return urlPath.substr(6).split('/');
+}
+
+// Determines whether path is a file
+function isFile(pathname) {
+    return pathname.split('/').pop().indexOf('.') > -1;
+}
+
+// Determines whether path is a directory
+function isDir(pathname) { return !isFile(pathname); }
```
Comment (humphd) [inline src/editor/src/routes.js]: All this `isDir` and `isFile` stuff should happen using `fs.stat` since you can't really know anything just with a filename.

File: src/editor/src/routes.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        // @ts-ignore
+        workbox.routing.registerRoute(
+            editRegex,
+            async ({ url, event }) => {
+                const paths = splitURL(url.pathname);
+
+                let folder = file = redirectURL = "";
+
+                // If URL contains a file
+                if (isFile(paths[paths.length - 1])) {
```
Comment (humphd) [inline src/editor/src/routes.js]: Change all this to use something like I showed you with `fs.stat(path, function(err, stats) {...}` and you can use `stats.isFile()` and `stats.isDirectory()`.

File: .gitignore (hunk 1 above)
Comment (twigz20) [reply .gitignore]: That file got somehow auto-generated exactly like that.

Comment (humphd) [reply .gitignore]: Can you remove it then?  It makes no sense to me.

Comment (twigz20) [reply .gitignore]: Will do.

File: src/editor/src/routes.js (hunk 6 above)
Comment (humphd) [reply src/editor/src/routes.js]: If you run `npm run build-webserver` it will overwrite the broken docs `index.html` and then you can use my old stuff for testing.

**Comment by twigz20** [comment]
@humphd I've made the necessary changes, however, I'm not able to test if the `fs.stats` method works.

File: src/editor/src/routes.js [hunk 7]
```diff
@@ -19,22 +14,27 @@ export default {
...
-                    file = paths[paths.length - 1];
-                    paths.pop();
-                    folder = paths.join('/');
-                    redirectURL = `${url.origin}/editor?folder=${folder}&file=${file}`;
-                }
-                else {
-                    folder = paths.join('/');
-                    redirectURL = `${url.origin}/editor?folder=${folder}`;
-                }
+                fs.stat(path, (err, stats) => {
```
Comment (humphd) [inline src/editor/src/routes.js]: Becuase `fs.stat()` is an async function, you have to wait on that callback before you run the code on line 38 below.  You need something like this:

```js
// Untested Code!!!!
fs.stat(path, (err, stats) => {
    if(err) {
        redirectURL = ...;
    } else {
        if(stats.isDirectory()) {
            ...
        } else {
            ...
        }
    }

    Response.redirect(redirectURL);
});
```

**Comment by twigz20** [comment]
@humphd Found a slight issue with uglify. Not sure if it needs to be updated or not. However, it fails here on `src/editor/dist/routes.js`
```
import fs from '../../lib/fs';
const sh = new fs.Shell();
```
with this error
```
Running "uglify:dist" (uglify) task
{ SyntaxError: Unexpected token: name (fs)
    at JS_Parse_Error.get (eval at <anonymous> (D:\Programming\WebDev\DPS911\Dev\next\src\editor\node_modules\uglify-js\tools\node.js:27:1), <anonymous>:86:23)dules\uglify-js\tools\node.js:27:1), <anonymous>:86:23)
    at formatError (util.js:640:16)
    at formatValue (util.js:544:18)
    at inspect (util.js:324:10)
    at format (util.js:191:12)
    at Console.log (console.js:130:21)                                                                  s:150:17
    at D:\Programming\WebDev\DPS911\Dev\next\src\editor\node_modules\grunt-contrib-uglify\tasks\uglify.jasks                                                                                                    glify\tasks\uglify.js:76:16)
\uglify.js:150:17                                                                                       \task.js:264:15)
    at D:\Programming\WebDev\DPS911\Dev\next\src\editor\node_modules\grunt-contrib-uglify\tunt-contrib-uasks\uglify.js:150:17
    at Array.forEach (<anonymous>)                                                         unt\lib\grunt    at Object.<anonymous> (D:\Programming\WebDev\DPS911\Dev\next\src\editor\node_modules\grunt-contrib-uglify\tasks\uglify.js:76:16)
    at Object.<anonymous> (D:\Programming\WebDev\DPS911\Dev\next\src\editor\node_modules\grunt\lib\grunt\task.js:264:15)
  message: 'Unexpected token: name (fs)',
  filename: 'routes.js',
  line: 1,
  col: 7,
  pos: 7 }
>> Uglifying source dist/routes.js failed.
```

**Comment by humphd** [comment]
It's due to using `import` without bundling.  Don't build that file in the editor build, build it as part of the parcel build with the web server.

**Comment by twigz20** [comment]
@humphd Alright, I think I'm good!

File: src/editor/src/routes.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        // @ts-ignore
+        workbox.routing.registerRoute(
+            editRegex,
+            async ({ url, event }) => {
+                const path = url.pathname.match(editRegex)[1];
+                const paths = splitURL(url.pathname);
+
+                let folder = file = redirectURL = "";
+                
+                // Untested Code!!!!
```
Comment (dsych) [inline src/editor/src/routes.js]: Please verify whether this code works and remove the comment.

File: tools/build.js [hunk 9]
```diff
@@ -13,6 +13,7 @@ mkdirp('dist')
     .then(() => commandRunner('npm run build-editor'))
     .then(() => mkdirp('dist/editor'))
     .then(() => copy('./src/editor/dist/**', './dist/editor'))
+    .then(() => commandRunner('parcel build src/editor/src/routes.js --out-dir dist/editor'))
```
Comment (dsych) [inline tools/build.js]: You don't need to explicitly build `routes.js` since they are included in the `service-worker.js`

File: src/editor/src/.editor.html [hunk 10]
```diff
@@ -0,0 +1,150 @@
+<!doctype html>
```
Comment (dsych) [inline src/editor/src/.editor.html]: You do we have both `.editor.html` and `editor.html`?

File: src/editor/src/routes.js (hunk 8 above)
Comment (twigz20) [reply src/editor/src/routes.js]: Cannot Test out the routes since the service worker cannot load with the current build. Once that is solved i'll be able to test.

**Comment by twigz20** [comment]
TODO: File a bug to rework the code to use `Filer.Path` instead of string manipulation.

**Comment by twigz20** [comment]
Hey @humphd @Fatehsandhu @mmBabol @dsych I'd like you guys to take a look at this.

**Comment by twigz20** [comment]
@dsych I've made the necessary changes. Please take a look.

**Comment by humphd** [comment]
@twigz20 can we get this landed soon?  I want to iterate on the editor, and can't without this.

**Comment by twigz20** [comment]
@humphd Should be good now.

**Comment by humphd** [comment]
@twigz20 can you fix your conflicts so we can merge?

**Comment by twigz20** [comment]
@humphd I'm confused? What conflicts?
screenshot_3

**Comment by twigz20** [comment]
@humphd I'm confused? What conflicts?
screenshot_18

**Comment by humphd** [comment]
I have `Rebase and merge` set, and you don't, that's all.

screen shot 2018-06-18 at 1 16 01 pm

**Comment by humphd** [comment]
So is this good to go then?  If so, let's squash and merge it!

**Comment by twigz20** [comment]
I believe so!

**Comment by humphd** [comment]
Merged.  Thanks!
